server/aws-lsp-codewhisperer/src/language-server/utilities/telemetryUtils.ts

```diff
@@ -0,0 +1,108 @@
+import { IdeCategory, UserContext } from '../../client/token/codewhispererbearertokenclient'
+
+const USER_AGENT_PREFIX = 'AWS-Language-Servers'
+
+export const getUserAgent = (initializeParams: InitializeParams, serverInfo?: ServerInfo): string => {
+    const format = (s: string) => s.replace(/\s/g, '-')
+
+    const items: String[] = []
+
+    items.push(USER_AGENT_PREFIX)
+
+    // Fields specific to runtime artifact
+    if (serverInfo?.name) {
+        serverInfo.version
+            ? items.push(`${format(serverInfo.name)}/${serverInfo.version}`)
+            : items.push(format(serverInfo.name))
+    }
+
+    // Compute client-specific suffix
+    // Missing required data fields are replaced with 'UNKNOWN' token
+    // Whitespaces in product.name and platform.name are replaced to '-'
+    if (initializeParams?.initializationOptions?.aws) {
+        const { clientInfo } = initializeParams?.initializationOptions?.aws
+        const { extension } = clientInfo || {}
+
+        if (extension) {
+            items.push(`${extension.name ? format(extension.name) : 'UNKNOWN'}/${extension.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+
+        if (clientInfo?.clientId) {
+            items.push(`ClientId/${clientInfo?.clientId}`)
+        }
+    } else {
+        // Default to standard InitializeParams.clientInfo if no custom aws.clientInfo is set
+        const { clientInfo } = initializeParams || {}
+        if (clientInfo) {
+            items.push(`${clientInfo.name ? format(clientInfo.name) : 'UNKNOWN'}/${clientInfo.version || 'UNKNOWN'}`)
+        }
+    }
+
+    return items.join(' ')
+}
+
+const IDE_CATEGORY_MAP: { [key: string]: IdeCategory } = {
+    'AmazonQ-For-VSCode': 'VSCODE',
+    'Amazon-Q-For-JetBrains': 'JETBRAINS',
+    'AmazonQ-For-Eclipse': 'ECLIPSE',
+    'AWS-Toolkit-For-VisualStudio': 'VISUAL_STUDIO',
+}
+
+const mapClientNameToIdeCategory = (clientName: string): string | undefined => {
+    return IDE_CATEGORY_MAP[clientName]
+}
+
+// Use InitializeParams.initializationOptions.aws.clientInfo.extension to derive IDE Category from calling client
+// https://github.com/aws/language-server-runtimes/blob/main/runtimes/protocol/lsp.ts#L60-L69
+const getIdeCategory = (initializeParams: InitializeParams) => {
+    let ideCategory
+    if (initializeParams.initializationOptions?.aws.clientInfo?.extension?.name) {
+        ideCategory = mapClientNameToIdeCategory(initializeParams.initializationOptions.aws.clientInfo.extension.name)
+    }
+
+    return ideCategory || 'UNKNOWN'
+}
+
+// Map result from https://nodejs.org/api/process.html#process_process_platform to expected Operating system
+const getOperatingSystem = () => {
+    // Node.js environment
+    if (typeof process !== 'undefined') {
```

Comment:
Would injecting this from runtime as part of `Runtime` be an option?

Reply:
that's an idea, we can make it part of Runtime feature we pass https://github.com/aws/language-server-runtimes/blob/main/runtimes/server-interface/runtime.ts